### PR TITLE
[ci] Skip Arrow tests on AppVeyor, use Intel macOS runners, upgrade to XCode 14.3 on macOS jobs, disable MacOS MPI jobs

### PR DESCRIPTION
**Pull request by borchero**
# Motivation

Fix failing CI on `main`.

This PR does two things:

- Skip Arrow tests in AppVeyor builds using an old Visual Studio version. The idea that `pyarrow` only works with newer versions of Visual Studio comes from [this comment](https://github.com/conda-forge/arrow-cpp-feedstock/issues/1374#issuecomment-2073656246) in the `pyarrow` feedstock:
  > You'll need at least VS2019 to successfully link the DLLs we're producing here
- Bump the version of Xcode used in MacOS GitHub runners (`macOS-latest` seems to have been updated recently).

### CODE REVIEW

**Comment by borchero** [comment]
This is still WIP, I just want to had CI running.

**Comment by borchero** [comment]
At least for compiler MSVC, using Visual Studio 2019 _would_ fix the issue. I'm also trying an older `pyarrow` build for completeness.

**Comment by jameslamb** [comment]
Thanks for investigating this! In case it helps, I searched this error message:

```text
  File "C:\Miniconda3-x64\envs\test-env\Scripts\pytest-script.py", line 9 in <module>
Windows fatal exception: code 0xc0000139
```

And found this relevant-looking discussion that's also about `pytest` + math packages + Windows: https://github.com/numba/numba/issues/8223.

----

re: some of the comments you got on https://github.com/conda-forge/arrow-cpp-feedstock/issues/1374 ...

here's the main script that runs for Appveyor jobs in this repo: https://github.com/microsoft/LightGBM/blob/master/.ci/test_r_package_windows.ps1. You can add any commands you want (like `conda info` or `conda env list`) and push a commit here to get a new build and new logs.

You should also be able to log in with your GitHub account at https://ci.appveyor.com/project/guolinke/lightgbm/history and cancel / restart any builds. If that doesn't work for you, let me know and we can get you that access.

**Comment by borchero** [comment]
All right, two observations:

- CI seems to work with Visual Studio 2019
- Something seems to have happened in https://github.com/conda-forge/arrow-cpp-feedstock/pull/1370 that started causing the issue (`pyarrow 15.0.2` build `3` works and CI starts failing with build `4` which _should_ be a build-only change)

File: tests/python_package_test/test_arrow.py [hunk 1]
```diff
@@ -132,7 +141,10 @@ def assert_datasets_equal(tmp_path: Path, lhs: lgb.Dataset, rhs: lgb.Dataset):
     ("arrow_table_fn", "dataset_params"),
     [  # Use lambda functions here to minimize memory consumption
         (lambda: generate_simple_arrow_table(), dummy_dataset_params()),
-        (lambda: generate_simple_arrow_table(empty_chunks=True), dummy_dataset_params()),
+        (
+            lambda: generate_simple_arrow_table(empty_chunks=True),
+            dummy_dataset_params(),
+        ),
```
Comment (jameslamb) [inline tests/python_package_test/test_arrow.py:147]: If the current state of this branch passes, don't bother reverting this... let's merge and unblock CI.

But if you do need to push another commit, could you please revert these style-only changes? They seem unrelated.

File: .ci/setup.sh [hunk 2]
```diff
@@ -17,7 +17,7 @@ if [[ $OS_NAME == "macos" ]]; then
             sudo xcode-select -s /Applications/Xcode_11.7.app/Contents/Developer || exit 1
         fi
     else  # gcc
-        sudo xcode-select -s /Applications/Xcode_14.1.app/Contents/Developer || exit 1
+        sudo xcode-select -s /Applications/Xcode_14.3.app/Contents/Developer || exit 1
```
Comment (jameslamb) [inline .ci/setup.sh:20]: Thanks for fixing this too!

**Review by jameslamb — APPROVED** [review]
Thank you SO MUCH for working through these issues! I really really appreciate it.

It looks to me like this is working great... I see the arrow tests running where we expect them to ([e.g. on Azure DevOps](https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build/results?buildId=16152&view=logs&j=275189f9-c769-596a-7ef9-49fb48a9ab70&t=3a9e7a4a-04e6-52a0-67ea-6e8f6cfda74f)) and being skipped on AppVeyor ([build link](https://ci.appveyor.com/project/guolinke/lightgbm/builds/49684413/job/l40g2bhp6ccsiyau)).

In the future when you work through these issues that are blocking all CI, feel free to kill any CI jobs that are blocking you. For example, #6426 came in while you were working on this, and I suspect you might have waited the 30-ish minutes for its AppVeyor jobs to run. Just wanted to say that it's totally fine to do that. My approach to that is like "those other PRs aren't going to be mergeable until I fix CI anyway, so they'll have to re-run anyway".

**Comment by borchero** [comment]
Since CI still doesn't work: I noticed that, very recently, `macos-latest` runners were upgraded from MacOS 12 to MacOS 14 **and** now use Apple Silicon M1 chips (i.e. an `arm` architecture), see [docs](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners/about-github-hosted-runners#standard-github-hosted-runners-for-public-repositories). This means, however, that many tests won't work on `macos-latest` (e.g. Python 3.7 is already EOL and not available for Apple Silicon on conda-forge). I am downgrading to `macos-13` now and we can go back to `macos-latest` in the future.

File: tests/python_package_test/test_arrow.py (hunk 1 above)
Comment (borchero) [reply tests/python_package_test/test_arrow.py]: Sorry, this was unintentional, got rid of this.

File: .github/workflows/r_package.yml [hunk 3]
```diff
@@ -3,11 +3,11 @@ name: R-package
 on:
   push:
     branches:
-    - master
+      - master
```
Comment (jameslamb) [inline .github/workflows/r_package.yml:6]: can you please revert all these unrelated formatting changes?

**Comment by jameslamb** [comment]
Thanks for the investigation! 100% agree with changest all the `macos-latest` to the latest intel version.

GitHub not having arm64 mac support for YEARS and then giving us just 3 months before making it the default is.... sigh 😫

File: .github/workflows/r_package.yml (hunk 3 above)
Comment (borchero) [reply .github/workflows/r_package.yml]: I really have to disable my format-on-save here 😬

**Comment by borchero** [comment]
Unfortunately, `cmake` fails to find the brew-installed OpenMPI package now although the `dylib` is located on the exact same path as before. Do you have an idea how to fix this @jameslamb?

**Comment by jameslamb** [comment]
I see this from `brew install open-mpi`.

```text
==> Pouring open-mpi--5.0.3.ventura.bottle.tar.gz
🍺  /usr/local/Cellar/open-mpi/5.0.3: 2,615 files, 74.4MB
```

And then this later on

```text
-- Could NOT find MPI_C (missing: MPI_C_WORKS) 
CMake Error at /usr/local/Cellar/cmake/3.29.2/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
-- Could NOT find MPI_CXX (missing: MPI_CXX_WORKS) 
  Could NOT find MPI (missing: MPI_C_FOUND MPI_CXX_FOUND)
-- Configuring incomplete, errors occurred!
Call Stack (most recent call first):
  /usr/local/Cellar/cmake/3.29.2/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:600 (_FPHSA_FAILURE_MESSAGE)
  /usr/local/Cellar/cmake/3.29.2/share/cmake/Modules/FindMPI.cmake:1837 (find_package_handle_standard_args)
  CMakeLists.txt:135 (find_package)
```

([build link](https://github.com/microsoft/LightGBM/actions/runs/8830205854/job/24242843629?pr=6425))

LightGBM is relying on the `FindMPI` script that ships with `CMake`. Here's the content of that script in case it helps: https://gitlab.kitware.com/cmake/cmake/-/blob/master/Modules/FindMPI.cmake.

I see you're trying the `CMAKE_PREFIX` fix mentioned at https://stackoverflow.com/a/71115879/3986677 ... at a minimum, I think that path needs to be into `/usr/local/Cellar/open-mpi/5.0.3`. But first, I recommend trying removing that `CMAKE_PREFIX` stuff and running this:

```shell
brew link open-mpi
```

ref: https://github.com/Homebrew/brew/blob/master/docs/FAQ.md#what-does-keg-only-mean

Sorry, will have more time to help later.

**Comment by jameslamb** [comment]
ugh that still didn't work.

I vote we just remove the MPI jobs on macOS for now. Delete the job definitions from the GitHub Actions config and document the need to get them working again in a new issue, linking back to this PR.

MPI jobs are working on Linux ([build link](https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build/results?buildId=16165&view=logs&j=11c46b0e-4322-5d4e-fafd-abe3157be7c4)), that's enough to cover most of the relevant code in the library.

Let's not let these MPI macOS tests block CI for the whole project.

**Comment by jameslamb** [comment]
@shiyu1994 can you please fix the VM where the CUDA tests run?

All CUDA CI jobs have been failing for the last 2 days. They hit something like this:

```text
Failed to restart docker.service: Connection timed out
See system logs and 'systemctl status docker.service' for details.
Error: Process completed with exit code 1.
```

([build link](https://github.com/microsoft/LightGBM/actions/runs/8827180505/job/24234040879))

Then hang until GitHub Actions kills them for hitting the 60 minute timeout.

File: .ci/setup.sh [hunk 4]
```diff
@@ -17,7 +17,9 @@ if [[ $OS_NAME == "macos" ]]; then
             sudo xcode-select -s /Applications/Xcode_11.7.app/Contents/Developer || exit 1
         fi
     else  # gcc
-        sudo xcode-select -s /Applications/Xcode_14.1.app/Contents/Developer || exit 1
+        # Check https://github.com/actions/runner-images/tree/main/images/macos for available
+        # versions of Xcode
```
Comment (jameslamb) [inline .ci/setup.sh:21]: forgot to say earlier... thanks very much for adding this comment on an otherwise kind of magical hard-coded string here 😁
